README.md
<div align="center">

# Video-GIF Converter

### Video to GIF and viceversa converter

[![NPM](https://nodei.co/npm/video-gif-converter.png?compact=true)](https://nodei.co/npm/video-gif-converter/)
<br />
[![](https://img.shields.io/npm/dt/video-gif-converter.svg?style=flat-square)](https://www.npmjs.com/package/video-gif-converter)

</div>

[![NPM Version](https://badgen.net/npm/v/video-gif-converter)](https://npmjs.org/package/video-gif-converter)
[![license](https://img.shields.io/github/license/pichillilorenzo/video-gif-converter)](/LICENSE)
[![Donate to this project using Paypal](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://www.paypal.me/LorenzoPichilli)

It uses [ffmpeg](https://ffmpeg.org/) and [gifski](https://github.com/ImageOptim/gifski) under the hood. They are both installed automatically as dependencies by this package.

## Getting started

To get started with this library, you need to install it and add it to your project.

### Installation

```bash
# npm
npm install video-gif-converter --save

# yarn
yarn add video-gif-converter
```

## Library Usage

API Reference available at [https://pichillilorenzo.github.io/video-gif-converter/](https://pichillilorenzo.github.io/video-gif-converter/).

```javascript
import * as path from 'path';
import {VideoGifConverter} from 'video-gif-converter';
// or
const path = require('path');
const {VideoGifConverter} = require('video-gif-converter');

const command = new VideoGifConverter({
  convertTo: VideoGifConverterConvertType.gif,
  output: path.join(__dirname, 'test', 'output'),
  gifski: true,
  quality: 100,
  startTime: 3,
  fps: 15,
  size: '640x?'
});
command.on('progress', progress => {
  console.log(progress);
});
// glob pattern
const files = path.join(__dirname, 'test', '**', '*.mp4');
const result = await command.run(files);
```

## CLI Usage

`<pattern>` represents a [glob pattern](https://www.npmjs.com/package/glob) used to specify input files to convert. Surround the glob pattern with quotes (example `'./test/**/*.mp4'`).

```
Usage: video-gif-converter [options] <pattern>

Examples: 
  video-gif-converter --gifski -q 100 -o './test/output' './test/video.mp4'
  video-gif-converter --use-palette -o './test/output' './test/video.mp4'
  video-gif-converter --fps 10 './test/**/*.mp4'
  video-gif-converter --convert-to video --format mp4 'test/**/*.gif'

Arguments:
  pattern                         Glob pattern of input files. Surround the glob pattern with quotes (example './test/**/*.mp4').

Options:
  -ct, --convert-to <type>        Convert to gif or video. (choices: "gif", "video", default: "gif")
  -o, --output <path>             Output folder path.
  -f, --format <format>           Output video format. Used and required only if --convert-to is video.
  -r, --fps <fps>                 Output framerate.
  -s, --size <size>               Output frame size (examples: 640x480, 640x?, 50%).
  -a, --aspect <aspect>           Output frame aspect ratio (examples: 4:3, 16:9).
  -g, --gifski                    Use gifski to generate GIF
  -q, --quality <1-100>           GIF output quality used by gifski
  -df, --delete-frames            Delete generated frames used by gifski (default: true)
  -vc, --video-codec <codec>      Video codec (examples: mpeg4, libx264).
  -vb, --video-bitrate <bitrate>  Sets the target video bitrate in kbps. The bitrate argument may be a number or a string with an optional k suffix.
  -vbk, --video-bitrate-constant  The constant argument specifies whether a constant bitrate should be enforced. (default: false)
  -d, --duration <seconds>        Forces ffmpeg to stop transcoding after a specific output duration. The time parameter may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
  -st, --start-time <seconds>     Seeks an input and only start decoding at given time offset. The time argument may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
  -l, --loop <times>              Set the number of times to loop the output. Use -1 for no loop, 0 for looping indefinitely (default).
  -up, --use-palette              Use ffmpeg palettegen and paletteuse filters to generate high quality GIF without gifski.
  -h, --help                      display help for command
```

## CLI Usage Example

```bash
video-gif-converter --gifski -q 100 -o './test/output' './test/**/*.mp4'
```

This code snippet shows how to put into action `video-gif-converter` to convert all `.mp4` video files into high quality GIFs using the [gifski](https://github.com/ImageOptim/gifski) encoder.

## Contributors

Any contribution is appreciated. You can get started with the steps below:

1. Fork [this repository](https://github.com/pichillilorenzo/video-gif-converter) (learn how to do this [here](https://help.github.com/articles/fork-a-repo)).

2. Clone the forked repository.

3. Make your changes and create a pull request ([learn how to do this](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)).

4. I will attend to your pull request and provide some feedback.

## License

This repository is licensed under the [ISC](LICENSE) License.


CHANGELOG.md
## 1.0.1

- Updated README

## 1.0.0

Initial release


test/lib.test.ts
import {describe, expect, test} from '@jest/globals';
import {VideoGifConverter, VideoGifConverterConvertType} from "../src";
import * as path from "path";
import * as fs from "fs";
import {spawnSync} from "child_process";

const rimraf = require('rimraf');

jest.setTimeout(70000);

const output = path.join(__dirname, 'output');

const videoGlobPattern = path.join(__dirname, '**', '*.mp4');
const videoFiles = [
  path.join(__dirname, 'video.mp4'),
  path.join(__dirname, 'video2.mp4')
];
const gifOutputFiles = [
  path.join(output, 'video.gif'),
  path.join(output, 'video2.gif')
];

// const gifGlobPattern = path.join(__dirname, '**', '*.gif');
const gifFiles = [
  path.join(__dirname, 'video.gif'),
  path.join(__dirname, 'video2.gif')
];
const videoOutputFiles = [
  path.join(output, 'video.mp4'),
  path.join(output, 'video2.mp4')
];

describe('video-gif-converter', () => {
  beforeEach(cb => {
    if (fs.existsSync(output)) {
      rimraf.sync(path.join(output, '**', '*'));
    }
    cb();
  });

  afterEach(cb => {
    if (fs.existsSync(output)) {
      rimraf.sync(path.join(output, '**', '*'));
    }
    cb();
  });

  test('basic usage', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output
    });
    const result = await command.run(videoFiles);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });

  test('basic usage run glob pattern', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output
    });
    const result = await command.run(videoGlobPattern);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });

  test('basic usage gifski', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output,
      gifski: true,
      quality: 75
    });
    const result = await command.run(videoFiles);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });

  test('use-palette', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output,
      usePalette: true
    });
    const result = await command.run(videoFiles);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });

  test('basic usage video', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.video,
      output,
      format: 'mp4',
    });
    const result = await command.run(gifFiles);
    videoOutputFiles.forEach(videoFile => {
      expect(fs.existsSync(videoFile)).toBeTruthy();
    });
    expect(result.completed).toBe(gifFiles.length);
  });

  test('basic usage error', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output
    });
    command.on('error', (err) => {
      expect(err).toBeDefined();
    });
    const result = await command.run([path.join(__dirname, 'video.mp4.not.existing')]);
    expect(fs.existsSync(path.join(output, 'video.mp4.not.existing'))).toBeFalsy();
    expect(result.failed).toBe(1);
  });

  test('progress event', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output
    });
    command.on('progress', progress => {
      expect(progress).toBeDefined();
    });
    const result = await command.run(videoFiles);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });

  test('end event', async () => {
    const command = new VideoGifConverter({
      convertTo: VideoGifConverterConvertType.gif,
      output
    });
    command.on('end', (result) => {
      expect(result).toBeDefined()
    });
    const result = await command.run(videoFiles);
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.completed).toBe(videoFiles.length);
  });
});

describe('video-gif-converter CLI', () => {
  beforeEach(cb => {
    if (fs.existsSync(output)) {
      rimraf.sync(path.join(output, '**', '*'));
    }
    cb();
  });

  afterEach(cb => {
    if (fs.existsSync(output)) {
      rimraf.sync(path.join(output, '**', '*'));
    }
    cb();
  });

  test('basic usage', async () => {
    const result = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '-o', output,
        videoGlobPattern
      ], {
        stdio: 'inherit'
      }
    );
    gifOutputFiles.forEach(gifFile => {
      expect(fs.existsSync(gifFile)).toBeTruthy();
    });
    expect(result.error).toBeUndefined();
  });
});


lib/cli.js
#!/usr/bin/env node
"use strict";Object.defineProperty(exports,"__esModule",{value:!0});const commander_1=require("commander"),glob_1=require("glob"),index_1=require("./index"),Bar=require("progress-barjs"),program=new commander_1.Command;async function convertFiles(t){const o={...program.opts()};if(t){const i=glob_1.glob.sync(`${t}`,{absolute:!0});console.log(`${i.length} file${i.length!==1?"s":""} found!`);const r=Bar({label:`Convert to .${o.convertTo===index_1.VideoGifConverterConvertType.gif?"gif":o.format}`,info:"Processing",total:i.length}),n=new index_1.VideoGifConverter(o);n.on("progress",()=>{r.tick("")}),console.log("Start processing...");const e=await n.run(i);e.failed>0&&console.error(`\u26A0\uFE0F ${e.failed} file${e.failed!==1?"s":""} failed. \u26A0\uFE0F`),e.completed>0&&(console.log(`\u{1F389} ${e.completed} file${e.completed!==1?"s have":" had"} been converted with success. \u{1F389}`),o.output&&console.log(`Files saved at: ${o.output}`))}else throw Error("Empty glob pattern!")}async function run(){const t=new commander_1.Option("-ct, --convert-to <type>","Convert to gif or video.").default(index_1.VideoGifConverterConvertType.gif).choices(Object.values(index_1.VideoGifConverterConvertType));t.required=!0,program.argument("<pattern>","Glob pattern of input files. Surround the glob pattern with quotes (example './test/**/*.mp4').").description(`Examples: 
  video-gif-converter --gifski -q 100 -o './test/output' './test/video.mp4'
  video-gif-converter --use-palette -o './test/output' './test/video.mp4'
  video-gif-converter --fps 10 './test/**/*.mp4'
  vvideo-gif-converter --convert-to video --format mp4 'test/**/*.gif'`).addOption(t).option("-o, --output <path>","Output folder path.").option("-f, --format <format>","Output video format. Used and required only if --convert-to is video.").option("-r, --fps <fps>","Output framerate.").option("-s, --size <size>","Output frame size (examples: 640x480, 640x?, 50%).").option("-a, --aspect <aspect>","Output frame aspect ratio (examples: 4:3, 16:9).").option("-g, --gifski","Use gifski to generate GIF").option("-q, --quality <1-100>","GIF output quality used by gifski").option("-df, --delete-frames","Delete generated frames used by gifski",!0).option("-vc, --video-codec <codec>","Video codec (examples: mpeg4, libx264).").option("-vb, --video-bitrate <bitrate>","Sets the target video bitrate in kbps. The bitrate argument may be a number or a string with an optional k suffix.").option("-vbk, --video-bitrate-constant","The constant argument specifies whether a constant bitrate should be enforced.",!1).option("-d, --duration <seconds>","Forces ffmpeg to stop transcoding after a specific output duration. The time parameter may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).").option("-st, --start-time <seconds>","Seeks an input and only start decoding at given time offset. The time argument may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).").option("-l, --loop <times>","Set the number of times to loop the output. Use -1 for no loop, 0 for looping indefinitely (default).").option("-up, --use-palette","Use ffmpeg palettegen and paletteuse filters to generate high quality GIF without gifski.").action(convertFiles),await program.parseAsync()}run();

lib/index.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0}),exports.VideoGifConverter=exports.VideoGifConverterConvertType=void 0;const tslib_1=require("tslib"),fs_1=require("fs"),fluent_ffmpeg_1=tslib_1.__importDefault(require("fluent-ffmpeg")),ffmpeg_static_1=tslib_1.__importDefault(require("ffmpeg-static")),gifski_command_1=require("gifski-command"),glob_1=require("glob"),events_1=tslib_1.__importDefault(require("events")),path=tslib_1.__importStar(require("path")),fs=tslib_1.__importStar(require("fs"));fluent_ffmpeg_1.default.setFfmpegPath(ffmpeg_static_1.default);var VideoGifConverterConvertType;(function(r){r.gif="gif",r.video="video"})(VideoGifConverterConvertType=exports.VideoGifConverterConvertType||(exports.VideoGifConverterConvertType={}));const ffmpegCommandRun=r=>new Promise((o,e)=>{r.on("error",(t,i,s)=>{o({err:t,stdout:i,sterr:s})}).on("end",(t,i,s)=>{o({err:t,stdout:i,sterr:s})}).run()});class VideoGifConverter extends events_1.default.EventEmitter{options;constructor(o){if(super(),this.options={convertTo:VideoGifConverterConvertType.gif,deleteFrames:!0,...o},this.options.convertTo===VideoGifConverterConvertType.video&&!this.options.format)throw Error("format is missing")}_buildCommand(o,e){const t=(0,fluent_ffmpeg_1.default)(o);if(this.options.fps&&t.fps(parseInt(this.options.fps.toString())),this.options.size&&t.size(this.options.size),this.options.aspect&&t.aspect(this.options.aspect),this.options.videoCodec&&t.videoCodec(this.options.videoCodec),this.options.videoBitrate&&t.videoBitrate(this.options.videoBitrate,this.options.videoBitrateConstant),this.options.startTime&&t.setStartTime(this.options.startTime),this.options.duration&&t.setDuration(this.options.duration),this.options.loop&&t.addOutputOption(["-loop",this.options.loop.toString()]),this.options.convertTo===VideoGifConverterConvertType.gif&&this.options.gifski){const i=this._buildOutputPath(o,`${e}.frame%04d.png`);t.output(i)}else if(this.options.convertTo===VideoGifConverterConvertType.gif){this.options.usePalette&&t.complexFilter(["split [o1] [o2];[o1] palettegen [p]; [o2] fifo [o3];[o3] [p] paletteuse"]);const i=this._buildOutputPath(o,"gif");t.toFormat("gif").output(i)}else{const i=this.options.format,s=this._buildOutputPath(o,i);t.toFormat(i).output(s)}return t}async run(o){typeof o=="string"&&(o=glob_1.glob.sync(o,{absolute:!0}));const e={completed:0,failed:0,total:o.length};this.options.output&&fs.mkdirSync(this.options.output,{recursive:!0});for(const t of o){const i=Date.now().toString(),s=this._buildCommand(t,i),n=await ffmpegCommandRun(s);if(n.err)console.error(n.err),e.failed++,this.emit("error",n.err,n.stdout,n.sterr);else if(this.options.convertTo===VideoGifConverterConvertType.gif&&this.options.gifski){const a=this._buildOutputPath(t,"gif"),u=this._buildOutputPath(t,`${i}.frame*.png`),p=await new gifski_command_1.GifskiCommand({output:a,frames:[u],quality:this.options.quality?parseInt(this.options.quality.toString()):void 0,width:9e3,repeat:this.options.loop}).run();if(p.err?(console.error(p.err),e.failed++,this.emit("error",p.err,p.stdout,p.stderr)):e.completed++,this.options.deleteFrames){const f=glob_1.glob.sync(u,{});for(const d of f)try{(0,fs_1.unlinkSync)(d)}catch(l){console.error(l)}}}else e.completed++;this.emit("progress",{...e,percent:(e.completed+e.failed)/e.total*100})}return this.emit("end",e),e}_buildOutputPath(o,e){const t=path.parse(o);return(this.options.output?path.join(this.options.output,t.name):path.join(t.dir,t.name))+`.${e}`}}exports.VideoGifConverter=VideoGifConverter;

src/index.ts
import {unlinkSync} from 'fs';
import FfmpegCommand from 'fluent-ffmpeg';
import ffmpegPath from "ffmpeg-static";
import {GifskiCommand} from "gifski-command";
import {glob} from "glob";
import events from "events";
import * as path from "path";
import * as fs from "fs";

FfmpegCommand.setFfmpegPath(ffmpegPath!);

/**
 * video-gif-converter conversion type
 */
export enum VideoGifConverterConvertType {
  gif = 'gif',
  video = 'video'
}

/**
 * video-gif-converter command options.
 */
export type VideoGifConverterOptions = {
  /**
   * Convert to gif or video.
   */
  convertTo: VideoGifConverterConvertType;

  /**
   * Output path.
   */
  output?: string;

  /**
   * Output video format. Used and required only if {@link convertTo} is {@link VideoGifConverterConvertType.video}.
   */
  format?: string;

  /**
   * Output framerate.
   */
  fps?: number;

  /**
   * Output frame size (examples: 640x480, 640x?, 50%).
   */
  size?: string;

  /**
   * Output frame aspect ratio (examples: 4:3, 16:9).
   */
  aspect?: string;

  /**
   * Use gifski to generate GIF.
   */
  gifski?: boolean;

  /**
   * GIF output quality used by gifski.
   */
  quality?: number;

  /**
   * Delete generated frames used by gifski.
   */
  deleteFrames?: boolean;

  /**
   * Video codec (examples: mpeg4, libx264).
   */
  videoCodec?: string;

  /**
   * Sets the target video bitrate in kbps. The bitrate argument may be a number or a string with an optional k suffix.
   */
  videoBitrate?: string;

  /**
   * The constant argument specifies whether a constant bitrate should be enforced.
   */
  videoBitrateConstant?: boolean;

  /**
   * Forces ffmpeg to stop transcoding after a specific output duration. The time parameter may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
   */
  startTime?: string | number;

  /**
   * Seeks an input and only start decoding at given time offset. The time argument may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
   */
  duration?: string | number;

  /**
   * The number of times to loop the output. Use -1 for no loop, 0 for looping indefinitely (default).
   */
  loop?: number;

  /**
   * Use ffmpeg palettegen and paletteuse filters to generate high quality GIF without gifski.
   */
  usePalette?: boolean;
}

/**
 * Result info of video-gif-converter command.
 */
export type VideoGifConverterResult = {
  /**
   * Number of conversion completed with success.
   */
  completed: number;

  /**
   * Number of failed conversion.
   */
  failed: number;

  /**
   * Total number of files to be converted.
   */
  total: number;
}

const ffmpegCommandRun = (ffmpegCommand: FfmpegCommand.FfmpegCommand) => {
  return new Promise<{ err?: Error; stdout?: string | Buffer; sterr?: string | Buffer; }>((resolve, _) => {
    ffmpegCommand
      .on('error', (err?: Error, stdout?: string | Buffer, sterr?: string | Buffer) => {
        resolve({err, stdout, sterr});
      })
      .on('end', (err?: Error, stdout?: string | Buffer, sterr?: string | Buffer) => {
        resolve({err, stdout, sterr});
      })
      .run();
  });
}

/**
 * video-gif-converter progress event interface.
 */
export interface VideoGifConverterProgress {
  /**
   * Total number of files completed with success.
   */
  completed: number;

  /**
   * Total number of files failed.
   */
  failed: number;

  /**
   * Total number of files.
   */
  total: number;

  /**
   * Progress percentage.
   */
  percent: number;
}

export declare interface VideoGifConverter {
  /**
   * Event called multiple times when video-gif-converter is running.
   *
   * @param event
   * @param listener
   */
  on(event: 'progress', listener: (name: VideoGifConverterProgress) => void): this;

  /**
   * Event called when video-gif-converter command is completed.
   *
   * @param event
   * @param listener
   */
  on(event: 'end', listener: (result: VideoGifConverterResult) => void): this;

  /**
   * Event called when video-gif-converter command emit an error.
   *
   * @param event
   * @param listener
   */
  on(event: 'error', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;
}

/**
 * video-gif-converter command.
 */
export class VideoGifConverter extends events.EventEmitter {
  options: VideoGifConverterOptions;

  constructor(options?: VideoGifConverterOptions) {
    super();
    this.options = {
      convertTo: VideoGifConverterConvertType.gif,
      deleteFrames: true,
      ...options
    };

    if (this.options.convertTo === VideoGifConverterConvertType.video) {
      if (!this.options.format) {
        throw Error('format is missing');
      }
    }
  }

  private _buildCommand(file: string, timestamp?: string): FfmpegCommand.FfmpegCommand {
    const ffmpegCommand = FfmpegCommand(file);
    if (this.options.fps) {
      ffmpegCommand.fps(parseInt(this.options.fps.toString()));
    }
    if (this.options.size) {
      ffmpegCommand.size(this.options.size);
    }
    if (this.options.aspect) {
      ffmpegCommand.aspect(this.options.aspect);
    }
    if (this.options.videoCodec) {
      ffmpegCommand.videoCodec(this.options.videoCodec);
    }
    if (this.options.videoBitrate) {
      ffmpegCommand.videoBitrate(this.options.videoBitrate, this.options.videoBitrateConstant);
    }
    if (this.options.startTime) {
      ffmpegCommand.setStartTime(this.options.startTime);
    }
    if (this.options.duration) {
      ffmpegCommand.setDuration(this.options.duration)
    }
    if (this.options.loop) {
      ffmpegCommand.addOutputOption(['-loop', this.options.loop.toString()]);
    }
    if (this.options.convertTo === VideoGifConverterConvertType.gif && this.options.gifski) {
      const output = this._buildOutputPath(file, `${timestamp}.frame%04d.png`);
      ffmpegCommand.output(output);
    } else {
      if (this.options.convertTo === VideoGifConverterConvertType.gif) {
        if (this.options.usePalette) {
          ffmpegCommand.complexFilter([
            `split [o1] [o2];[o1] palettegen [p]; [o2] fifo [o3];[o3] [p] paletteuse`
          ]);
        }
        const output = this._buildOutputPath(file, 'gif');
        ffmpegCommand
          .toFormat('gif')
          .output(output)
      } else {
        const format = this.options.format!;
        const output = this._buildOutputPath(file, format);
        ffmpegCommand
          .toFormat(format)
          .output(output)
      }
    }
    return ffmpegCommand;
  }

  /**
   * Run video-gif-converter command.
   *
   * @param files glob pattern or list of input files.
   */
  async run(files: string[] | string): Promise<VideoGifConverterResult> {
    if (typeof files === 'string') {
      files = glob.sync(files, {
        absolute: true
      });
    }

    const result: VideoGifConverterResult = {
      completed: 0,
      failed: 0,
      total: files.length
    };

    if (this.options.output) {
      fs.mkdirSync(this.options.output, {recursive: true});
    }

    for (const file of files) {
      const timestamp = Date.now().toString();
      const ffmpegCommand = this._buildCommand(file, timestamp);
      const ffmpegResult = await ffmpegCommandRun(ffmpegCommand);

      if (ffmpegResult.err) {
        console.error(ffmpegResult.err);
        result.failed++;
        this.emit('error', ffmpegResult.err, ffmpegResult.stdout, ffmpegResult.sterr);
      } else {
        if (this.options.convertTo === VideoGifConverterConvertType.gif && this.options.gifski) {
          const output = this._buildOutputPath(file, 'gif');
          const framesOutput = this._buildOutputPath(file, `${timestamp}.frame*.png`);
          const gifskiCommand = new GifskiCommand({
            output: output,
            frames: [framesOutput],
            quality: this.options.quality ? parseInt(this.options.quality.toString()) : undefined,
            width: 9000, // big enough to set automatic size based on ffmpeg scale option
            repeat: this.options.loop
          });
          const gifskiResult = await gifskiCommand.run();
          if (gifskiResult.err) {
            console.error(gifskiResult.err);
            result.failed++;
            this.emit('error', gifskiResult.err, gifskiResult.stdout, gifskiResult.stderr);
          } else {
            result.completed++;
          }
          if (this.options.deleteFrames) {
            const frames = glob.sync(framesOutput, {});
            for (const frame of frames) {
              try {
                unlinkSync(frame);
              } catch (e) {
                console.error(e);
              }
            }
          }
        } else {
          result.completed++;
        }
      }

      this.emit('progress', {
        ...result,
        percent: (result.completed + result.failed) / result.total * 100
      });
    }

    this.emit('end', result);

    return result;
  }

  private _buildOutputPath(file: string, ext: string) {
    const parsedPath = path.parse(file);
    return (this.options.output ?
        path.join(this.options.output, parsedPath.name) :
        path.join(parsedPath.dir, parsedPath.name)) +
      `.${ext}`;
  }
}


lib/index.d.ts
/// <reference types="node" />
/// <reference types="node" />
import events from "events";
/**
 * video-gif-converter conversion type
 */
export declare enum VideoGifConverterConvertType {
    gif = "gif",
    video = "video"
}
/**
 * video-gif-converter command options.
 */
export declare type VideoGifConverterOptions = {
    /**
     * Convert to gif or video.
     */
    convertTo: VideoGifConverterConvertType;
    /**
     * Output path.
     */
    output?: string;
    /**
     * Output video format. Used and required only if {@link convertTo} is {@link VideoGifConverterConvertType.video}.
     */
    format?: string;
    /**
     * Output framerate.
     */
    fps?: number;
    /**
     * Output frame size (examples: 640x480, 640x?, 50%).
     */
    size?: string;
    /**
     * Output frame aspect ratio (examples: 4:3, 16:9).
     */
    aspect?: string;
    /**
     * Use gifski to generate GIF.
     */
    gifski?: boolean;
    /**
     * GIF output quality used by gifski.
     */
    quality?: number;
    /**
     * Delete generated frames used by gifski.
     */
    deleteFrames?: boolean;
    /**
     * Video codec (examples: mpeg4, libx264).
     */
    videoCodec?: string;
    /**
     * Sets the target video bitrate in kbps. The bitrate argument may be a number or a string with an optional k suffix.
     */
    videoBitrate?: string;
    /**
     * The constant argument specifies whether a constant bitrate should be enforced.
     */
    videoBitrateConstant?: boolean;
    /**
     * Forces ffmpeg to stop transcoding after a specific output duration. The time parameter may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
     */
    startTime?: string | number;
    /**
     * Seeks an input and only start decoding at given time offset. The time argument may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).
     */
    duration?: string | number;
    /**
     * The number of times to loop the output. Use -1 for no loop, 0 for looping indefinitely (default).
     */
    loop?: number;
    /**
     * Use ffmpeg palettegen and paletteuse filters to generate high quality GIF without gifski.
     */
    usePalette?: boolean;
};
/**
 * Result info of video-gif-converter command.
 */
export declare type VideoGifConverterResult = {
    /**
     * Number of conversion completed with success.
     */
    completed: number;
    /**
     * Number of failed conversion.
     */
    failed: number;
    /**
     * Total number of files to be converted.
     */
    total: number;
};
/**
 * video-gif-converter progress event interface.
 */
export interface VideoGifConverterProgress {
    /**
     * Total number of files completed with success.
     */
    completed: number;
    /**
     * Total number of files failed.
     */
    failed: number;
    /**
     * Total number of files.
     */
    total: number;
    /**
     * Progress percentage.
     */
    percent: number;
}
export declare interface VideoGifConverter {
    /**
     * Event called multiple times when video-gif-converter is running.
     *
     * @param event
     * @param listener
     */
    on(event: 'progress', listener: (name: VideoGifConverterProgress) => void): this;
    /**
     * Event called when video-gif-converter command is completed.
     *
     * @param event
     * @param listener
     */
    on(event: 'end', listener: (result: VideoGifConverterResult) => void): this;
    /**
     * Event called when video-gif-converter command emit an error.
     *
     * @param event
     * @param listener
     */
    on(event: 'error', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;
}
/**
 * video-gif-converter command.
 */
export declare class VideoGifConverter extends events.EventEmitter {
    options: VideoGifConverterOptions;
    constructor(options?: VideoGifConverterOptions);
    private _buildCommand;
    /**
     * Run video-gif-converter command.
     *
     * @param files glob pattern or list of input files.
     */
    run(files: string[] | string): Promise<VideoGifConverterResult>;
    private _buildOutputPath;
}


src/cli.ts
#!/usr/bin/env node
import {Command, Option} from 'commander';
import {glob} from 'glob';
import {VideoGifConverterConvertType, VideoGifConverterOptions, VideoGifConverter} from './index';
const Bar = require('progress-barjs');

const program = new Command();

async function convertFiles(pattern: string) {
  const options: VideoGifConverterOptions = {
    ...program.opts(),
  };
  if (pattern) {
    const files = glob.sync(`${pattern}`, {
      absolute: true
    });

    console.log(`${files.length} file${files.length !== 1 ? 's' : ''} found!`);

    const bar = Bar({
      label: `Convert to .${options.convertTo === VideoGifConverterConvertType.gif ? 'gif' : options.format}`,
      info: 'Processing',
      total: files.length,
    });

    const command = new VideoGifConverter(options);
    command.on('progress', () => {
      bar.tick('');
    });
    console.log(`Start processing...`);
    const result = await command.run(files);
    if (result.failed > 0) {
      console.error(`⚠️ ${result.failed} file${result.failed !== 1 ? 's' : ''} failed. ⚠️`);
    }
    if (result.completed > 0) {
      console.log(`🎉 ${result.completed} file${result.completed !== 1 ? 's have' : ' had'} been converted with success. 🎉`);
      if (options.output) {
        console.log(`Files saved at: ${options.output}`);
      }
    }
  } else {
    throw Error('Empty glob pattern!');
  }
}

async function run() {
  const convertTypeOption = new Option(
      '-ct, --convert-to <type>',
      'Convert to gif or video.'
    )
    .default(VideoGifConverterConvertType.gif)
    .choices(Object.values(VideoGifConverterConvertType));
  convertTypeOption.required = true;

  program
    .argument('<pattern>', `Glob pattern of input files. Surround the glob pattern with quotes (example './test/**/*.mp4').`)
    .description(`Examples: 
  video-gif-converter --gifski -q 100 -o './test/output' './test/video.mp4'
  video-gif-converter --use-palette -o './test/output' './test/video.mp4'
  video-gif-converter --fps 10 './test/**/*.mp4'
  vvideo-gif-converter --convert-to video --format mp4 'test/**/*.gif'`)
    .addOption(convertTypeOption)
    .option('-o, --output <path>', 'Output folder path.')
    .option('-f, --format <format>', 'Output video format. Used and required only if --convert-to is video.')
    .option('-r, --fps <fps>', 'Output framerate.')
    .option('-s, --size <size>', 'Output frame size (examples: 640x480, 640x?, 50%).')
    .option('-a, --aspect <aspect>', 'Output frame aspect ratio (examples: 4:3, 16:9).')
    .option('-g, --gifski', 'Use gifski to generate GIF')
    .option('-q, --quality <1-100>', 'GIF output quality used by gifski')
    .option('-df, --delete-frames', 'Delete generated frames used by gifski', true)
    .option('-vc, --video-codec <codec>', 'Video codec (examples: mpeg4, libx264).')
    .option('-vb, --video-bitrate <bitrate>', 'Sets the target video bitrate in kbps. The bitrate argument may be a number or a string with an optional k suffix.')
    .option('-vbk, --video-bitrate-constant', 'The constant argument specifies whether a constant bitrate should be enforced.', false)
    .option('-d, --duration <seconds>', 'Forces ffmpeg to stop transcoding after a specific output duration. The time parameter may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).')
    .option('-st, --start-time <seconds>', 'Seeks an input and only start decoding at given time offset. The time argument may be a number (in seconds) or a timestamp string (with format [[hh:]mm:]ss[.xxx]).')
    .option('-l, --loop <times>', 'Set the number of times to loop the output. Use -1 for no loop, 0 for looping indefinitely (default).')
    .option('-up, --use-palette', 'Use ffmpeg palettegen and paletteuse filters to generate high quality GIF without gifski.')
    .action(convertFiles);
  await program.parseAsync();
}

run();
